scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -334,11 +334,218 @@ object OptimizationRules {
       Binary(recurse(hj.left), rightLit,
         { (ltp: TypedPipe[(K, V)], rtp: TypedPipe[(K, V2)]) =>
           rtp match {
-            case ReduceStepPipe(hg: HashJoinable[K, V2]) =>
+            case ReduceStepPipe(hg: HashJoinable[K @unchecked, V2 @unchecked]) =>
               HashCoGroup(ltp, hg, joiner)
             case otherwise =>
               HashCoGroup(ltp, IdentityReduce(ordK, otherwise, None, Nil), joiner)
           }
         })
     }
+
+
+  /////////////////////////////
+  //
+  // Here are some actual rules for simplifying TypedPipes
+  //
+  /////////////////////////////
+
+  object ComposeFlatMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(FlatMapped(in, fn0), fn1) =>
+        FlatMapped(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+      case FlatMapValues(FlatMapValues(in, fn0), fn1) =>
+        FlatMapValues(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+    }
+  }
+
+  object ComposeMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case Mapped(Mapped(in, fn0), fn1) =>
+        Mapped(in, ComposedMapFn(fn0, fn1))
+      case MapValues(MapValues(in, fn0), fn1) =>
+        MapValues(in, ComposedMapFn(fn0, fn1))
+    }
+  }
+
+  object ComposeFilter extends Rule[TypedPipe] {
+    def apply[T](on: Dag[TypedPipe]) = {
+      // scala can't type check this, so we hold its hand:
+      // case Filter(Filter(in, fn0), fn1) =>
+      //   Some(Filter(in, ComposedFilterFn(fn0, fn1)))
+      case f@Filter(_, _) =>
+        def go[A](f: Filter[A]): Option[TypedPipe[A]] =
+          f.input match {
+            case f1: Filter[a] =>
+              Some(Filter[a](f1.input, ComposedFilterFn(f.fn, f.fn)))
+            case _ => None
+          }
+        go(f)
+      case FilterKeys(FilterKeys(in, fn0), fn1) =>
+        Some(FilterKeys(in, ComposedFilterFn(fn0, fn1)))
+      case _ => None
+    }
+  }
+
+  object ComposeWithOnComplete extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case WithOnComplete(WithOnComplete(pipe, fn0), fn1) =>
+        WithOnComplete(pipe, ComposedOnComplete(fn0, fn1))
+    }
+  }
+
+  object RemoveDuplicateForceFork extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case ForceToDisk(ForceToDisk(t)) => ForceToDisk(t)
+      case ForceToDisk(Fork(t)) => ForceToDisk(t)
+      case Fork(Fork(t)) => Fork(t)
+      case Fork(ForceToDisk(t)) => ForceToDisk(t)
+    }
+  }
+
+  /**
+   * We ignore .group if there are is no setting of reducers
+   *
+   * This is arguably not a great idea, but scalding has always
+   * done it to minimize accidental map-reduce steps
+   */
+  object IgnoreNoOpGroup extends PartialRule[TypedPipe] {
```

Comment:
This is something that has tripped a lot of first time users of scalding. Should we add a warning log line when this rule is being applied to say that we're ignoring the group?

Reply:
Adding a warning log line is not a bad idea in general. Not sure the right place to do that. I kind of hate to do it in the optimizer, since you might not even be running the result. Let me think about it.